compiler_opt/rl/env.py

```diff
@@ -219,6 +222,135 @@ def _reward_fn(a: float, b: float) -> float:
+  communicate to) also opens the pipe. If that doesn't happen, we time out.
+  Afterwards, `write` ops shouldn't block.
+  """
+  opened = threading.Event()
+  timed_out = threading.Event()
+
+  # start a thread that waits for `open` to unblock. If it doesn't, we open the
+  # fifo ourselves just to unblock.
+  def _timeout_thread():
+    if opened.wait(timeout):
+      logging.debug('[timeout thread] writer opened successfully')
+      return
+    timed_out.set()
+    logging.debug('[timeout thread] writer failed to open')
+    with open(filename, 'rb'):
+      pass
+
+  waiter = threading.Thread(target=_timeout_thread)
+  waiter.start()
+  try:
+    with io.BufferedWriter(io.FileIO(filename, 'wb')) as writer_pipe:
+      if not timed_out.is_set():
+        opened.set()
+        yield writer_pipe
+  finally:
+    waiter.join()
+    if timed_out.is_set():
+      # it's possible that the timeout thread timed out but also the other
+      # process finally opened the pipe and thus the `writer_pipe` is
+      # functional, but at the end we still raise TimeoutError. We accept that
+      # right now.
+      raise TimeoutError('write pipe open')
+
+
+@contextlib.contextmanager
+def open_read_pipe(filename: str, *, timeout: float):
+  """Open the read pipe, with a timeout governing the open and each read.
+
+  Just like in the writer case, assuming we're opening a fifo pipe, the open
+  operation will block until the other party opens the pipe. Then, because this
+  is a reader, each read operation (and variations - readline, etc) can block,
+  but no more than the provided timeout.
+  """
+
+  # wrap the underlying io.RawIOBase such that we poll before attempting to read
+  def _wrap_raw_io(obj: io.RawIOBase):
+
+    def _get_polling_wrapper(wrapped_method):
+
+      def _replacement(*args, **kwargs):
+        name = wrapped_method.__name__
+        logging.debug('ReaderWithTimeout is asked to %s', name)
+        (r, _, _) = select.select([obj], [], [], timeout)
+        if r:
+          logging.debug('ReaderWithTimeout %s should be unblocked', name)
+          result = wrapped_method(*args, **kwargs)
+          logging.debug('ReaderWithTimeout %s completed', name)
+          return result
+        logging.info('ReaderWithTimeout timed out waiting to %s', name)
+        raise TimeoutError('timed out reading')
+
+      return _replacement
+
+    # pylint: disable=protected-access
+    obj._orig_read = obj.read
+    obj._orig_readline = obj.readline
+    obj._orig_readinto = obj.readinto
+    obj._orig_readall = obj.readall
+
+    obj.read = _get_polling_wrapper(obj._orig_read)
+    obj.readline = _get_polling_wrapper(obj._orig_readline)
+    obj.readinto = _get_polling_wrapper(obj._orig_readinto)
+    obj.readall = _get_polling_wrapper(obj._orig_readall)
+    # pylint: enable=protected-access
+    return obj
+
+  opened = threading.Event()
+  timed_out = threading.Event()
+
+  # same idea as in the writer case - unblock the `open`
```

Comment:
Looks like this isn't worth factoring out due to all the differences in logging messages?